Applications/Spire/Source/Charting/ChartView.cpp

```diff
@@ -460,20 +482,80 @@ void ChartView::resizeEvent(QResizeEvent* event) {
 }
 
 void ChartView::showEvent(QShowEvent* event) {
-  if(m_top_left.m_x == ChartValue() && m_top_left.m_y == ChartValue() &&
-      m_bottom_right.m_x == ChartValue() &&
-      m_bottom_right.m_y == ChartValue()) {
+  if(m_region == Region{}) {
     auto current_time = boost::posix_time::second_clock::local_time();
     auto bottom_right = ChartPoint(ChartValue(current_time),
       ChartValue(Nexus::Money(0)));
     auto top_left = ChartPoint(
       ChartValue(current_time - boost::posix_time::hours(1)),
       ChartValue(Nexus::Money(1)));
-    set_region(top_left, bottom_right);
+    set_region({top_left, bottom_right});
   }
   QWidget::showEvent(event);
 }
 
+ChartView::GapInfo ChartView::update_gaps(std::vector<ChartView::Gap>& gaps,
+    std::vector<Candlestick>& candlesticks, ChartValue start) {
+  auto gap_info = GapInfo{0, ChartValue()};
+  for(auto& candlestick : candlesticks) {
+    auto end = candlestick.GetStart();
+    if(start != end) {
+      gaps.push_back({start, end});
+      gap_info.total_gaps_value += end - start;
+      ++gap_info.gap_count;
+    }
+    start = candlestick.GetEnd();
+  }
+  return gap_info;
+}
+
+QtPromise<ChartView::LoadedData> ChartView::load_data(
+    QtPromise<std::vector<Candlestick>>& promise, LoadedData data,
+    ChartModel* model) {
+  return promise.then([=] (auto result) mutable {
+    auto new_candlesticks = std::move(result.Get());
+    if(!new_candlesticks.empty() && !data.m_candlesticks.empty() &&
+        new_candlesticks.front().GetStart() <=
+        data.m_candlesticks.back().GetStart()) {
+      auto last = std::find_if(new_candlesticks.begin(),
+        new_candlesticks.end(), [&] (const auto& candlestick) {
+          return candlestick.GetStart() >=
+            data.m_candlesticks.back().GetStart();
+        });
+      auto b = new_candlesticks.begin();
+      if(last != new_candlesticks.end()) {
+        last++;
+      }
+      new_candlesticks.erase(b, last);
+    }
+    auto last = [&] {
+      if(data.m_candlesticks.empty() && new_candlesticks.empty()) {
+        return ChartValue();
+      } else if(data.m_candlesticks.empty()) {
+        return new_candlesticks.front().GetStart();
+      }
+      return data.m_candlesticks.back().GetEnd();
+    }();
+    auto gap_info = update_gaps(data.m_gaps, new_candlesticks, last);
+    data.m_candlesticks.insert(data.m_candlesticks.end(),
+      new_candlesticks.begin(), new_candlesticks.end());
+    data.m_current_x += std::min(static_cast<int>(std::ceil((
```

Comment:
If I understand this right I have to keep the first argument (the call to ceil()) as a variable to prevent it from being destroyed before min() can return it?

Reply:
Yeah the whole thing is retarded, we'll fix all of these in one shot by writing our own min/max functions.